#Bike Sales baseline classsification model



###Problem statement
###Objective
Creating a classification model that can forecast which gender will drink softdrink more the other.

##Hypothesis

trained model for gender classification repeated the most will have lower accuracy compared to the test model

##Import necessary libraries

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle
from collections import Counter
from sklearn.metrics import accuracy_score, classification_report

##Load data

In [2]:
df=pd.read_csv('gender classification.csv')

##data exploration

In [3]:
#display the first 5 dataset
df.head()

Favorite Color Favorite Music Genre Favorite Beverage Favorite Soft Drink  \
0           Cool                 Rock             Vodka          7UP/Sprite   
1        Neutral              Hip hop             Vodka     Coca Cola/Pepsi   
2           Warm                 Rock              Wine     Coca Cola/Pepsi   
3           Warm     Folk/Traditional           Whiskey               Fanta   
4           Cool                 Rock             Vodka     Coca Cola/Pepsi   

  Gender  
0      F  
1      F  
2      F  
3      F  
4      F

In [4]:
#display the number of rows and columns in the data
df.shape

(66, 5)

In [5]:
#display columns in the data
df.columns

Index(['Favorite Color', 'Favorite Music Genre', 'Favorite Beverage',
       'Favorite Soft Drink', 'Gender'],
      dtype='object')

In [6]:
#summary of data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Favorite Color        66 non-null     object
 1   Favorite Music Genre  66 non-null     object
 2   Favorite Beverage     66 non-null     object
 3   Favorite Soft Drink   66 non-null     object
 4   Gender                66 non-null     object
dtypes: object(5)
memory usage: 2.7+ KB


In [7]:
#statistical summary of categorical values
df.describe(include='O')

Favorite Color Favorite Music Genre Favorite Beverage  \
count              66                   66                66   
unique              3                    7                 6   
top              Cool                 Rock     Doesn't drink   
freq               37                   19                14   

       Favorite Soft Drink Gender  
count                   66     66  
unique                   4      2  
top        Coca Cola/Pepsi      F  
freq                    32     33

In [8]:
#identify datatypes
df.dtypes

Favorite Color          object
Favorite Music Genre    object
Favorite Beverage       object
Favorite Soft Drink     object
Gender                  object
dtype: object

In [9]:
#check null values
df.isnull().sum()

Favorite Color          0
Favorite Music Genre    0
Favorite Beverage       0
Favorite Soft Drink     0
Gender                  0
dtype: int64

###create baseline model

In [10]:
# Shuffling the Dataset
df = shuffle(df, random_state = 42)

In [11]:
df.head()

Favorite Color Favorite Music Genre Favorite Beverage Favorite Soft Drink  \
54           Cool           Electronic     Doesn't drink               Fanta   
62           Cool              Hip hop              Beer     Coca Cola/Pepsi   
0            Cool                 Rock             Vodka          7UP/Sprite   
45           Cool              Hip hop              Beer     Coca Cola/Pepsi   
5            Warm           Jazz/Blues     Doesn't drink               Fanta   

   Gender  
54      M  
62      M  
0       F  
45      M  
5       F

In [12]:
#actual value
X=df['Favorite Soft Drink']

In [13]:
target_column = 'Gender'

# Extract the target values
y = df[target_column]

In [14]:
# Calculate the mode of the target variable
gender_mode = y.mode()[0]
print(gender_mode)

F


In [15]:
predictions = [gender_mode] * len(y)

In [16]:
predictions

['F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F',
 'F']

In [17]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df.drop(target_column, axis=1), y, test_size=0.2, random_state=42)

In [18]:
predictions = [gender_mode] * len(y)

accuracy = accuracy_score(y, predictions)

print(f"accuracy: {accuracy}")

accuracy: 0.5


In [19]:
# Calculate mean values for the training set
train_classification_mode = y_train.mode()[0]

In [20]:
# Create a baseline model that predicts the mean of the training set for all samples
train_predictions = [train_classification_mode] * len(y_train)
train_accuracy = accuracy_score(y_train, train_predictions)
print(f"training accuracy: {train_accuracy}")

training accuracy: 0.5192307692307693


In [21]:
# Calculate precision, recall, f1-score for training
report = classification_report(y_train, train_predictions)
print(f'Gender Classification Report:\n{report}')

Gender Classification Report:
              precision    recall  f1-score   support

           F       0.52      1.00      0.68        27
           M       0.00      0.00      0.00        25

    accuracy                           0.52        52
   macro avg       0.26      0.50      0.34        52
weighted avg       0.27      0.52      0.35        52



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [22]:
df['train_classification_mode'] = train_classification_mode

In [23]:
df.head()

Favorite Color Favorite Music Genre Favorite Beverage Favorite Soft Drink  \
54           Cool           Electronic     Doesn't drink               Fanta   
62           Cool              Hip hop              Beer     Coca Cola/Pepsi   
0            Cool                 Rock             Vodka          7UP/Sprite   
45           Cool              Hip hop              Beer     Coca Cola/Pepsi   
5            Warm           Jazz/Blues     Doesn't drink               Fanta   

   Gender train_classification_mode  
54      M                         F  
62      M                         F  
0       F                         F  
45      M                         F  
5       F                         F

In [24]:
# Replace the y_test values with the mean of the training set
y_test_mode= pd.Series([train_classification_mode] * len(y_test), index=y_test.index)

y_test = pd.DataFrame(y_test)
y_test['train_classification_mode'] = y_test_mode
y_test

Gender train_classification_mode
55      M                         F
7       F                         F
54      M                         F
49      M                         F
63      M                         F
42      M                         F
3       F                         F
41      M                         F
51      M                         F
31      F                         F
15      F                         F
30      F                         F
46      M                         F
13      F                         F

In [25]:
# Calculate accuracy,precision, recall, f1-score for test dataset in training set
test_accuracy = accuracy_score(y_test['Gender'], y_test['train_classification_mode'])
report = classification_report(y_test['Gender'], y_test['train_classification_mode'])
print(f"test accuracy: {train_accuracy}")
print(f'Gender Classification Report:\n{report}')

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


test accuracy: 0.5192307692307693
Gender Classification Report:
              precision    recall  f1-score   support

           F       0.43      1.00      0.60         6
           M       0.00      0.00      0.00         8

    accuracy                           0.43        14
   macro avg       0.21      0.50      0.30        14
weighted avg       0.18      0.43      0.26        14



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


##Conclusions

Since the female class is the most prevalent in this dataset, we shall provide insight based on its results.

The predictive power of the model is similar to random guessing inside the training set, as indicated by the accuracy of the 80% trained value, which is roughly 52%. Recall, F1-scores of 68%, supporters of 27 and precision at the chance level for each class are also included in the classification report.

This accuracy level suggests that the model's predictions on the testing set are less accurate than random guessing, since the testing data shows an accuracy of roughly 43%. The classification report, which shows metrics well below chance, further demonstrates the bad performance across a number of classes.

##Recommendation

To develop a more reliable and precise classification model, it is essential to carefully reexamine the model's structure and possibly gather new data or rephrase the issue.